### ISSUE TITLE
Regression in .NET 9-RC2: FileLoadException in Microsoft.Extensions.ApiDescription.Server with Swashbuckle

### ISSUE THREAD

**Issue by bkoelman** [dotnet/aspnetcore#58379]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

When using the package `Swashbuckle.AspNetCore` with `Microsoft.Extensions.ApiDescription.Server`, the project fails to compile with the following error:
```
1>  dotnet "C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\../tools/dotnet-getdocument.dll" --assembly "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\bin\Debug\net9.0\SwashbuckleGenAtCompileTimeDemo.dll" --file-list "obj\SwashbuckleGenAtCompileTimeDemo.OpenApiFiles.cache" --framework ".NETCoreApp,Version=v9.0" --output "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\GeneratedSwagger\net9.0" --project "SwashbuckleGenAtCompileTimeDemo" --assets-file "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\obj\project.assets.json" --platform "AnyCPU"
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error : System.IO.FileLoadException: Could not load file or assembly 'Microsoft.OpenApi, Version=1.6.17.0, Culture=neutral, PublicKeyToken=3f5743946376f042'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error : File name: 'Microsoft.OpenApi, Version=1.6.17.0, Culture=neutral, PublicKeyToken=3f5743946376f042'
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :  ---> System.IO.FileLoadException: Could not load file or assembly 'C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\tools\net9.0\Microsoft.OpenApi.dll'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error : File name: 'C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\tools\net9.0\Microsoft.OpenApi.dll'
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommand.<>c__DisplayClass8_0.<Execute>b__1(AssemblyLoadContext loadContext, AssemblyName assemblyName)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at System.Runtime.Loader.AssemblyLoadContext.GetFirstResolvedAssemblyFromResolvingEvent(AssemblyName assemblyName)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingResolvingEvent(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker..cctor()
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.GetDocuments(IServiceProvider services)
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.Process()
1>C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\Microsoft.Extensions.ApiDescription.Server.targets(68,5): error MSB3073: The command "dotnet "C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\build\../tools/dotnet-getdocument.dll" --assembly "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\bin\Debug\net9.0\SwashbuckleGenAtCompileTimeDemo.dll" --file-list "obj\SwashbuckleGenAtCompileTimeDemo.OpenApiFiles.cache" --framework ".NETCoreApp,Version=v9.0" --output "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\GeneratedSwagger\net9.0" --project "SwashbuckleGenAtCompileTimeDemo" --assets-file "D:\Bart\Source\Projects\SwashbuckleGenAtCompileTimeDemo\obj\project.assets.json" --platform "AnyCPU" " exited with code 11.
1>Done building project "SwashbuckleGenAtCompileTimeDemo.csproj" -- FAILED.
```

### Expected Behavior

No compile error and generation of the OpenAPI JSON file at compile time in subdirectory `GeneratedSwagger\net9.0`.

### Steps To Reproduce

1. Create a new ASP.NET Web API project, targeting .NET 9 with "Enable OpenAPI support" and "Use controllers" checked
2. In the project file, replace the contents inside `<ItemGroup>` with the following:
    ```xml
    <PackageReference Include="Microsoft.Extensions.ApiDescription.Server" Version="9.0.0-rc.2.24474.3" PrivateAssets="all" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.8.1" />
    ```
3. Replace the contents of `Program.cs` with the following code:
    ```c#
    var builder = WebApplication.CreateBuilder(args);
    
    // Add services to the container.
    
    builder.Services.AddControllers();
    // Learn more about configuring OpenAPI at https://aka.ms/aspnet/openapi
    //builder.Services.AddOpenApi();
    builder.Services.AddSwaggerGen();
    
    var app = builder.Build();
    
    // Configure the HTTP request pipeline.
    if (app.Environment.IsDevelopment())
    {
        //app.MapOpenApi();
        app.UseSwagger(options => options.RouteTemplate = "openapi/{documentName}.json");
    }
    
    app.UseHttpsRedirection();
    
    app.UseAuthorization();
    
    app.MapControllers();
    
    app.Run();
    ```
4. Run the project and browse to `/openapi/v1.json`, which shows the OpenAPI JSON for the WeatherForecast API endpoint.
5. In the project file, add the following lines inside `<PropertyGroup>`:
    ```xml
    <OpenApiGenerateDocumentsOnBuild>true</OpenApiGenerateDocumentsOnBuild>
    <OpenApiDocumentsDirectory>GeneratedSwagger\$(TargetFramework)</OpenApiDocumentsDirectory>
    ```
6. Compile the project and observe the error mentioned above.

### Exceptions (if any)

```
System.IO.FileLoadException: Could not load file or assembly 'Microsoft.OpenApi, Version=1.6.17.0, Culture=neutral, PublicKeyToken=3f5743946376f042'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
File name: 'Microsoft.OpenApi, Version=1.6.17.0, Culture=neutral, PublicKeyToken=3f5743946376f042'
 ---> System.IO.FileLoadException: Could not load file or assembly 'C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\tools\net9.0\Microsoft.OpenApi.dll'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
File name: 'C:\Users\Bart\.nuget\packages\microsoft.extensions.apidescription.server\9.0.0-rc.2.24474.3\tools\net9.0\Microsoft.OpenApi.dll'
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommand.<>c__DisplayClass8_0.<Execute>b__1(AssemblyLoadContext loadContext, AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.GetFirstResolvedAssemblyFromResolvingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingResolvingEvent(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
   at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker..cctor()
   at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.GetDocuments(IServiceProvider services)
   at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.Process()
```

### .NET Version

9.0.100-rc.2.24474.11

### Anything else?

Visual Studio version: 17.12.0 Preview 2.1

```
> dotnet --info
.NET SDK:
 Version:           9.0.100-rc.2.24474.11
 Commit:            315e1305db
 Workload version:  9.0.100-manifests.82e6a096
 MSBuild version:   17.12.0-preview-24473-03+fea15fbd1

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-rc.2.24474.11\

.NET workloads installed:
 [aspire]
   Installation Source: VS 17.12.35323.107
   Manifest Version:    8.2.0/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.2.0\WorkloadManifest.json
   Install Type:              Msi

Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0-rc.2.24473.5
  Architecture: x64
  Commit:       990ebf52fc

.NET SDKs installed:
  9.0.100-rc.1.24452.12 [C:\Program Files\dotnet\sdk]
  9.0.100-rc.2.24474.11 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0-rc.1.24452.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0-rc.2.24474.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0-rc.1.24431.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0-rc.2.24473.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0-rc.1.24452.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0-rc.2.24474.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by martincostello** [comment]
This is probably because Microsoft.AspNetCore.OpenApi depends on version [1.6.17](https://github.com/dotnet/aspnetcore/blob/59260392fc4f31eba3ed6b97bc1b4375eeb8e0b9/eng/Versions.props#L333) of Microsoft.OpenApi, yet the current latest version of Swashbuckle.AspNetCore depends on [1.6.14](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/349ba38301c35ddf118cf401d48bcaa4d3d66284/Directory.Packages.props#L24-L25).

There'll be a new version of Swashbuckle.AspNetCore next month that depends on [whatever the latest version is once .NET 9 ships](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/pull/3007/files#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156).

A point-in-time fix would be to have the two reference the exact same version, but that isn't feasible long-term.

I'm not sure if there's a solution that would make this be a non-issue indefinitely, but I imagine the issue will go away if your app explicitly references (or uses [transitive pinning](https://learn.microsoft.com/nuget/consume-packages/central-package-management#transitive-pinning) if you use Central Package Management) Microsoft.OpenApi 1.6.17 or later.

**Comment by bkoelman** [comment]
@martincostello Thanks for the quick response, the helpful links, and the suggested workaround. I can confirm that adding `<PackageReference Include="Microsoft.OpenApi" Version="1.6.17" />` resolves the issue.

This unblocks my preliminary testing, so I'm fine closing this issue, as the needed changes for .NET 9 are already tracked elsewhere.

**Comment by 0xced** [comment]
Today I got the exact same issue when upgrading `Microsoft.Extensions.ApiDescription.Server` from version 10.0.11 to 10.0.12. I had to add an explicit package reference to `Microsoft.OpenApi` version 2.12.2 to resolve the issue.

Error below to help people find it.

```
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  : System.TypeInitializationException: The type initializer for 'Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker' threw an exception.
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :  ---> System.IO.FileLoadException: Could not load file or assembly 'Microsoft.OpenApi, Version=2.12.0.0, Culture=neutral, PublicKeyToken=3f5743946376f042'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  : File name: 'Microsoft.OpenApi, Version=2.12.0.0, Culture=neutral, PublicKeyToken=3f5743946376f042'
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :  ---> System.IO.FileLoadException: Could not load file or assembly '~/.nuget/packages/microsoft.extensions.apidescription.server/10.0.12/tools/net10.0/Microsoft.OpenApi.dll'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  : File name: '~/.nuget/packages/microsoft.extensions.apidescription.server/10.0.12/tools/net10.0/Microsoft.OpenApi.dll'
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommand.<>c__DisplayClass8_0.<Execute>b__1(AssemblyLoadContext loadContext, AssemblyName assemblyName)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at System.Runtime.Loader.AssemblyLoadContext.GetFirstResolvedAssemblyFromResolvingEvent(AssemblyName assemblyName)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker..cctor()
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    --- End of inner exception stack trace ---
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.GetDocuments(IServiceProvider services)
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error  :    at Microsoft.Extensions.ApiDescription.Tool.Commands.GetDocumentCommandWorker.Process()
33>Microsoft.Extensions.ApiDescription.Server.targets(67,5): Error MSB3073 : The command "dotnet ~/.nuget/packages/microsoft.extensions.apidescription.server/10.0.12/build/../tools/dotnet-getdocument.dll --assembly [...] --platform "AnyCPU" " exited with code 11.
```
